tests/test-validate-app-receipt.js

```diff
@@ -0,0 +1,208 @@
+
+    var testReceipt = makeReceipt(null, {
+      typ: 'test-receipt',
+      iss: 'https://payments-alt.allizom.org',
+      verify: 'https://payments-alt.allizom.org/developers/test-receipt/',
+    });
+    helper.appSelf.receipts = [testReceipt];
+
+    var validator = new helper.ReceiptValidator({
+      verifyUrl: new RegExp(
+        'https://payments-alt\\.allizom\\.org/developers/test-receipt/'),
+    });
+
+    fxpay.validateAppReceipt(function(error, productInfo) {
+      if (!error) {
+        assert.equal(productInfo.receiptInfo.status, 'ok');
+      }
+      done(error);
+    });
+
+    validator.finish();
+  });
+
+  it('fails when no receipt is present', function(done) {
+    helper.appSelf.receipts = [];
+
+    var validator = new helper.ReceiptValidator();
+
+    fxpay.validateAppReceipt(function(error, productInfo) {
+      assert.equal(error, 'APP_RECEIPT_MISSING');
+      assert.equal(typeof productInfo, 'object');
+      done();
+    });
+
+    validator.finish();
+  });
+
+  it('fails when mozApps is null', function(done) {
+    fxpay.configure({mozApps: null});
+
+    fxpay.validateAppReceipt(function(error, productInfo) {
+      assert.equal(error, 'PAY_PLATFORM_UNAVAILABLE');
+      assert.equal(typeof productInfo, 'object');
+      done();
+    });
+
+  });
+
+  it('fails when appSelf is null', function(done) {
+    helper.appSelf.result = null;
+
+    fxpay.validateAppReceipt(function(error, productInfo) {
+      assert.equal(error, 'PAY_PLATFORM_UNAVAILABLE');
+      assert.equal(typeof productInfo, 'object');
+      done();
+    });
+
+    helper.appSelf.onsuccess();
+  });
+
+  it('fails when multiple receipts are installed', function(done) {
+    helper.appSelf.receipts = [makeReceipt(), makeReceipt()];
+
+    fxpay.validateAppReceipt(function(error, productInfo) {
+      assert.equal(error, 'NOT_IMPLEMENTED_ERROR');
+      assert.equal(typeof productInfo, 'object');
+      done();
+    });
+
+    helper.appSelf.onsuccess();
+  });
+
+  it('fails when receipt is malformed', function(done) {
+    helper.appSelf.receipts = ['^%%%$$$$garbage'];
+
+    var validator = new helper.ReceiptValidator();
+
+    fxpay.validateAppReceipt(function(error, productInfo) {
+      assert.equal(error, 'APP_RECEIPT_MISSING');
+      assert.equal(typeof productInfo, 'object');
+      done();
+    });
+
+    validator.finish();
+  });
+
+  it('ignores in-app receipts', function(done) {
+    var appId = '1234';
+    var appReceipt = makeReceipt({storedata: 'id=' + appId});
+    var inAppProductReceipt = makeReceipt({storedata: 'id=555&inapp_id=234'});
+    helper.appSelf.receipts = [
+      appReceipt,
+      inAppProductReceipt,
+    ];
+
+    var validator = new helper.ReceiptValidator();
+
+    fxpay.validateAppReceipt(function(error, productInfo) {
+      assert.equal(productInfo.productId, appId);
+      done(error);
+    });
+
+    validator.finish();
+  });
+
+
+  function makeReceipt(overrides, receiptData) {
```

Comment:
The way I was thinking about it originally is that it's more important to read test code so I put all the test code at the top. The helper functions are important only after you find yourself digging into a test and wanting to learn more.